<a href="https://colab.research.google.com/github/lim-jr/ESAA/blob/main/%EC%B7%A8%ED%95%A9_%EB%AF%B8%EB%8B%88%ED%94%8C%EC%A0%9D2_%EB%AF%BC%EA%B2%BD%2C%ED%98%84%EC%84%9C%EC%A0%84%EC%B2%98%EB%A6%AC%EB%B0%98%EC%98%81.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. 데이터셋 불러오기

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
import numpy as np
import pandas as pd 

In [24]:
train = pd.read_csv("/content/drive/MyDrive/PyCo_ESAA/데이터셋/train 지역변수 추가.csv")
test = pd.read_csv("/content/drive/MyDrive/PyCo_ESAA/데이터셋/test 지역변수 추가.csv")

In [4]:
test.head(3)

inst_id  OC     sido  sgg    openDate  bedCount          instkind  \
0        2 NaN  incheon  139  19981125.0     300.0  general_hospital   
1        5 NaN     jeju  149  20160309.0      44.0          hospital   
2        6 NaN  jeonnam  103  19890427.0     276.0  general_hospital   

       revenue1    salescost1          sga1  ...  liquidLiabilities2  \
0  6.682486e+10  0.000000e+00  6.565709e+10  ...        5.068443e+10   
1  3.495758e+10  0.000000e+00  3.259270e+10  ...        4.209828e+10   
2  2.326031e+10  2.542571e+09  2.308749e+10  ...        2.777589e+10   

     shortLoan2  NCLiabilities2     longLoan2     netAsset2      surplus2  \
0  3.714334e+10    4.720000e+09  4.690000e+09  1.608540e+10  8.944587e+09   
1  2.420000e+10    2.521009e+10  1.830000e+10  3.789135e+09  0.000000e+00   
2  2.182278e+10    0.000000e+00  0.000000e+00  0.000000e+00  1.638540e+10   

   employee1  employee2  ownerChange  hospital_count  
0      693.0      693.0         same             1.0  
1      379.0      371.0         same             1.0  
2        NaN        NaN          NaN             3.0  

[3 rows x 59 columns]

In [5]:
train.head(3)

inst_id    OC       sido  sgg  openDate  bedCount          instkind  \
0        1  open  choongnam   73  20071228     175.0  nursing_hospital   
1        3  open  gyeongnam   32  19970401     410.0  general_hospital   
2        4  open   gyeonggi   89  20161228     468.0  nursing_hospital   

       revenue1   salescost1          sga1  ...  liquidLiabilities2  \
0  4.217530e+09          0.0  3.961135e+09  ...         222876855.0   
1           NaN          NaN           NaN  ...                 NaN   
2  1.004522e+09  515483669.0  4.472197e+08  ...                 0.0   

   shortLoan2  NCLiabilities2    longLoan2     netAsset2      surplus2  \
0         0.0     536116887.0  390000000.0  2.619290e+09  1.271224e+09   
1         NaN             NaN          NaN           NaN           NaN   
2         0.0             0.0          0.0  0.000000e+00  0.000000e+00   

   employee1  employee2  ownerChange  hospital_count  
0       62.0       64.0         same            14.0  
1      801.0      813.0         same             5.0  
2      234.0        1.0         same            17.0  

[3 rows x 59 columns]

In [6]:
train.describe()

inst_id         sgg      openDate    bedCount      revenue1  \
count  301.000000  301.000000  3.010000e+02  296.000000  2.930000e+02   
mean   219.056478   81.039867  2.005013e+07  145.709459  1.288175e+10   
std    121.234869   50.969714  8.893815e+04  118.923890  2.043543e+10   
min      1.000000    1.000000  1.978012e+07    0.000000  0.000000e+00   
25%    112.000000   37.000000  2.001102e+07   52.750000  3.252112e+09   
50%    230.000000   75.000000  2.007113e+07  136.500000  5.524218e+09   
75%    321.000000  123.000000  2.011102e+07  193.000000  1.274839e+10   
max    428.000000  178.000000  2.017061e+07  656.000000  1.510000e+11   

         salescost1          sga1       salary1          noi1          noe1  \
count  2.930000e+02  2.930000e+02  2.930000e+02  2.930000e+02  2.930000e+02   
mean   2.014903e+09  1.033244e+10  5.654115e+09  2.696151e+08  5.115870e+08   
std    7.460271e+09  1.493886e+10  8.083343e+09  8.023701e+08  1.060379e+09   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    0.000000e+00  2.758201e+09  1.626053e+09  8.217133e+06  8.013395e+07   
50%    2.104105e+08  4.684074e+09  2.659892e+09  4.363764e+07  1.831965e+08   
75%    9.105278e+08  1.077876e+10  6.363400e+09  2.050331e+08  4.203330e+08   
max    9.850332e+10  1.030000e+11  6.403559e+10  9.144171e+09  8.686380e+09   

       ...         debt2  liquidLiabilities2    shortLoan2  NCLiabilities2  \
count  ...  2.930000e+02        2.930000e+02  2.930000e+02    2.930000e+02   
mean   ...  8.146026e+09        3.860584e+09  1.510050e+09    4.471247e+09   
std    ...  1.255800e+10        6.797242e+09  2.953412e+09    7.659580e+09   
min    ...  0.000000e+00        0.000000e+00  0.000000e+00    0.000000e+00   
25%    ...  1.283220e+09        2.855741e+08  0.000000e+00    2.557878e+07   
50%    ...  3.784553e+09        1.454050e+09  8.542549e+07    1.969747e+09   
75%    ...  8.465053e+09        4.364714e+09  1.567967e+09    4.905441e+09   
max    ...  8.508858e+10        6.846878e+10  1.759375e+10    5.150388e+10   

          longLoan2     netAsset2      surplus2    employee1    employee2  \
count  2.930000e+02  2.930000e+02  2.930000e+02   291.000000   288.000000   
mean   2.709979e+09  5.273919e+09  9.786279e+08   142.546392   134.326389   
std    4.564001e+09  1.081259e+10  4.688798e+09   160.191073   151.061786   
min    0.000000e+00 -2.014780e+10 -2.781507e+09     0.000000     0.000000   
25%    0.000000e+00  1.017573e+09  0.000000e+00    53.500000    53.750000   
50%    1.100000e+09  2.894970e+09  0.000000e+00    80.000000    79.000000   
75%    3.360000e+09  5.370285e+09  1.014447e+08   181.500000   170.000000   
max    3.256147e+10  1.250000e+11  6.852730e+10  1200.000000  1200.000000   

       hospital_count  
count      299.000000  
mean        10.230769  
std          6.221061  
min          1.000000  
25%          5.000000  
50%         10.000000  
75%         15.500000  
max         20.000000  

[8 rows x 55 columns]

### # 변수설명
inst_id - 각 파일에서의 병원 고유 번호   
OC – 영업/폐업 분류, 2018년 폐업은 2017년 폐업으로 간주함    
sido – 병원의 광역 지역 정보    
sgg – 병원의 시군구 자료     
openDate – 병원 설립일    
bedCount - 병원이 갖추고 있는 병상의 수       
instkind – 병원, 의원, 요양병원, 한의원, 종합병원 등 병원의 종류    
·        종합병원 : 입원환자 100명 이상 수용 가능    
·        병원 : 입원 환자 30명 이상 100명 미만 수용 가능     
·        의원 : 입원 환자 30명 이하 수용 가능      
·        한방 병원(한의원) : 침술과 한약으로 치료하는 의료 기관.    
revenue1 – 매출액 / salescost1 – 매출원가      
sga1 - 판매비와 관리비 / salary1 – 급여   
noi1 – 영업외수익 / noe1 – 영업외비용      
Interest1 – 이자비용 / ctax1 – 법인세비용     
Profit1 – 당기순이익 / liquidAsset1 – 유동자산      
quickAsset1 – 당좌자산 / receivableS1 - 미수금(단기)       
inventoryAsset1 – 재고자산 / nonCAsset1 – 비유동자산     
tanAsset1 – 유형자산 / OnonCAsset1 - 기타 비유동자산       
receivableL1 – 장기미수금 / debt1 – 부채총계      
liquidLiabilities1 – 유동부채 / shortLoan1 – 단기차입금    
NCLiabilities1 – 비유동부채 / longLoan1 – 장기차입금    
netAsset1 – 순자산총계 / surplus1 – 이익잉여금    
employee1 – 고용한 총 직원의 수     
ownerChange – 대표자의 변동 

# 2. EDA 및 전처리

In [7]:
train.isnull().sum()

inst_id                0
OC                     0
sido                   0
sgg                    0
openDate               0
bedCount               5
instkind               1
revenue1               8
salescost1             8
sga1                   8
salary1                8
noi1                   8
noe1                   8
interest1              8
ctax1                  8
profit1                8
liquidAsset1           8
quickAsset1            8
receivableS1           8
inventoryAsset1        8
nonCAsset1             8
tanAsset1              8
OnonCAsset1            8
receivableL1           8
debt1                  8
liquidLiabilities1     8
shortLoan1             8
NCLiabilities1         8
longLoan1              8
netAsset1              8
surplus1               8
revenue2               8
salescost2             8
sga2                   8
salary2                8
noi2                   8
noe2                   8
interest2              8
ctax2                  8
profit2                8


In [8]:
test.isnull().sum()

inst_id                 0
OC                    127
sido                    0
sgg                     0
openDate                1
bedCount                8
instkind                2
revenue1                2
salescost1              2
sga1                    2
salary1                 2
noi1                    2
noe1                    2
interest1               2
ctax1                   2
profit1                 2
liquidAsset1            2
quickAsset1             2
receivableS1            2
inventoryAsset1         2
nonCAsset1              2
tanAsset1               2
OnonCAsset1             2
receivableL1            2
debt1                   2
liquidLiabilities1      2
shortLoan1              2
NCLiabilities1          2
longLoan1               2
netAsset1               2
surplus1                2
revenue2                2
salescost2              2
sga2                    2
salary2                 2
noi2                    2
noe2                    2
interest2               2
ctax2       

### (1) 레이블 데이터 : 인코딩

In [25]:
from sklearn.preprocessing import LabelEncoder

# OC : 'close': 0, 'open': 1 처리해주기
print(train['OC'].value_counts()) #인코딩전 확인
encoder1 = LabelEncoder()
train['OC']=encoder1.fit_transform(train['OC'])
print(train['OC'].value_counts()) #인코딩후 확인

# 테스트데이터셋은 OC가 모두 NaN이므로 레이블인코딩 처리x

open      286
 close     15
Name: OC, dtype: int64
1    286
0     15
Name: OC, dtype: int64


### (2) 피처 데이터 : 결측치 처리 및 인코딩

#### (2)-1. ints_id, sgg

In [26]:
# inst_id : 인덱스 중복이므로 열 버리기
# sgg : sido와 정보가 겹치므로 열 버리기
train.drop(['inst_id','sgg'],axis=1,inplace=True)
test.drop(['inst_id','sgg'],axis=1,inplace=True)

#### (2)-2. 문자열 피처

In [27]:
## 문자열 피처에서 결측치가 존재할 경우, 'Not_sure'로 대체
# 문자열 피처중 instkind, ownerChange 에서 결측치 존재

train['instkind'] = train['instkind'].fillna('Not_sure')
test['instkind'] = test['instkind'].fillna('Not_sure')

train['ownerChange'] = train['ownerChange'].fillna('Not_sure')
test['ownerChange'] = test['ownerChange'].fillna('Not_sure')

In [28]:
## 문자열 피처 인코딩 처리

# sido는 레이블인코딩 (sido의 개별관측치가 많아 원핫인코딩을 하게 되면 변수개수가 너무많아지기 때문.)
from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()
train['sido'] = encoder.fit_transform(train['sido'])
test['sido'] = encoder.fit_transform(test['sido'])

# instkind, ownerChange는 원핫인코딩
train = pd.get_dummies(train)
test = pd.get_dummies(test)

In [29]:
## 확인
train.head(3)
# sido는 숫자형변수로, instkind, ownerChange는 더미변수로 잘 바뀜.

OC  sido  openDate  bedCount      revenue1   salescost1          sga1  \
0   1     2  20071228     175.0  4.217530e+09          0.0  3.961135e+09   
1   1     9  19970401     410.0           NaN          NaN           NaN   
2   1     8  20161228     468.0  1.004522e+09  515483669.0  4.472197e+08   

        salary1        noi1        noe1  ...  instkind_clinic  \
0  2.033835e+09  15652441.0  15236236.0  ...                0   
1           NaN         NaN         NaN  ...                0   
2  2.964023e+08     76156.0     30000.0  ...                0   

   instkind_dental_clinic  instkind_general_hospital  instkind_hospital  \
0                       0                          0                  0   
1                       0                          1                  0   
2                       0                          0                  0   

   instkind_nursing_hospital  instkind_traditional_clinic  \
0                          1                            0   
1                          0                            0   
2                          1                            0   

   instkind_traditional_hospital  ownerChange_Not_sure  ownerChange_change  \
0                              0                     0                   0   
1                              0                     0                   0   
2                              0                     0                   0   

   ownerChange_same  
0                 1  
1                 1  
2                 1  

[3 rows x 66 columns]

#### (2)-3. 숫자형 피처중 금융변수

In [30]:
## 숫자형 피처중 금융변수에서 결측치(NaN) 존재

# open&같은행번호에서 공통적으로 8개의 결측치가 나타난다. 
print(train[train['revenue1'].isnull()]['OC'])
print(train[train['shortLoan2'].isnull()]['OC'])

# 같은행번호에서 공통적으로 2개의 결측치가 나타난다.
print(test[test['revenue1'].isnull()])
print(test[test['shortLoan2'].isnull()])

1      1
14     1
32     1
55     1
60     1
248    1
257    1
258    1
Name: OC, dtype: int64
1      1
14     1
32     1
55     1
60     1
248    1
257    1
258    1
Name: OC, dtype: int64
    OC  sido    openDate  bedCount  revenue1  salescost1  sga1  salary1  noi1  \
15 NaN     0  19950330.0     226.0       NaN         NaN   NaN      NaN   NaN   
56 NaN    10  20121218.0     539.0       NaN         NaN   NaN      NaN   NaN   

    noe1  ...  instkind_Not_sure  instkind_clinic  instkind_general_hospital  \
15   NaN  ...                  0                0                          1   
56   NaN  ...                  0                0                          0   

    instkind_hospital  instkind_nursing_hospital  instkind_traditional_clinic  \
15                  0                          0                            0   
56                  0                          1                            0   

    instkind_traditional_hospital  ownerChange_Not_sure  ownerChange_change  \
15

In [31]:
#NAN인 관측치 처리(dropna)
fin_list=['revenue1', 'salescost1', 'sga1', 'salary1', 'noi1', 'noe1',
       'interest1', 'ctax1', 'profit1', 'liquidAsset1', 'quickAsset1',
       'receivableS1', 'inventoryAsset1', 'nonCAsset1', 'tanAsset1',
       'OnonCAsset1', 'receivableL1', 'debt1', 'liquidLiabilities1',
       'shortLoan1', 'NCLiabilities1', 'longLoan1', 'netAsset1', 'surplus1',
       'revenue2', 'salescost2', 'sga2', 'salary2', 'noi2', 'noe2',
       'interest2', 'ctax2', 'profit2', 'liquidAsset2', 'quickAsset2',
       'receivableS2', 'inventoryAsset2', 'nonCAsset2', 'tanAsset2',
       'OnonCAsset2', 'receivableL2', 'debt2', 'liquidLiabilities2',
       'shortLoan2', 'NCLiabilities2', 'longLoan2', 'netAsset2', 'surplus2']
train.dropna(subset=fin_list, axis=0,inplace=True)

#0인 관측치 처리
train[fin_list]=train[fin_list].replace(0 ,-9999)

In [32]:
## 확인
train.describe()
# 최소값이 -9999로 나타난다.

OC        sido      openDate    bedCount      revenue1  \
count  293.000000  293.000000  2.930000e+02  288.000000  2.930000e+02   
mean     0.948805    7.098976  2.005059e+07  143.663194  1.288175e+10   
std      0.220771    4.203769  8.854337e+04  118.062972  2.043543e+10   
min      0.000000    0.000000  1.978012e+07    0.000000 -9.999000e+03   
25%      1.000000    3.000000  2.001122e+07   52.750000  3.252112e+09   
50%      1.000000    8.000000  2.007120e+07  134.500000  5.524218e+09   
75%      1.000000   10.000000  2.011102e+07  186.750000  1.274839e+10   
max      1.000000   15.000000  2.017061e+07  656.000000  1.510000e+11   

         salescost1          sga1       salary1          noi1          noe1  \
count  2.930000e+02  2.930000e+02  2.930000e+02  2.930000e+02  2.930000e+02   
mean   2.014900e+09  1.033244e+10  5.654114e+09  2.696149e+08  5.115866e+08   
std    7.460272e+09  1.493886e+10  8.083343e+09  8.023702e+08  1.060379e+09   
min   -9.999000e+03 -9.999000e+03 -9.999000e+03 -9.999000e+03 -9.999000e+03   
25%   -9.999000e+03  2.758201e+09  1.626053e+09  8.217133e+06  8.013395e+07   
50%    2.104105e+08  4.684074e+09  2.659892e+09  4.363764e+07  1.831965e+08   
75%    9.105278e+08  1.077876e+10  6.363400e+09  2.050331e+08  4.203330e+08   
max    9.850332e+10  1.030000e+11  6.403559e+10  9.144171e+09  8.686380e+09   

       ...  instkind_clinic  instkind_dental_clinic  \
count  ...       293.000000              293.000000   
mean   ...         0.058020                0.003413   
std    ...         0.234182                0.058421   
min    ...         0.000000                0.000000   
25%    ...         0.000000                0.000000   
50%    ...         0.000000                0.000000   
75%    ...         0.000000                0.000000   
max    ...         1.000000                1.000000   

       instkind_general_hospital  instkind_hospital  \
count                 293.000000         293.000000   
mean                    0.112628           0.300341   
std                     0.316678           0.459191   
min                     0.000000           0.000000   
25%                     0.000000           0.000000   
50%                     0.000000           0.000000   
75%                     0.000000           1.000000   
max                     1.000000           1.000000   

       instkind_nursing_hospital  instkind_traditional_clinic  \
count                 293.000000                   293.000000   
mean                    0.481229                     0.013652   
std                     0.500502                     0.116239   
min                     0.000000                     0.000000   
25%                     0.000000                     0.000000   
50%                     0.000000                     0.000000   
75%                     1.000000                     0.000000   
max                     1.000000                     1.000000   

       instkind_traditional_hospital  ownerChange_Not_sure  \
count                     293.000000            293.000000   
mean                        0.027304              0.034130   
std                         0.163246              0.181873   
min                         0.000000              0.000000   
25%                         0.000000              0.000000   
50%                         0.000000              0.000000   
75%                         0.000000              0.000000   
max                         1.000000              1.000000   

       ownerChange_change  ownerChange_same  
count          293.000000        293.000000  
mean             0.143345          0.822526  
std              0.351024          0.382723  
min              0.000000          0.000000  
25%              0.000000          1.000000  
50%              0.000000          1.000000  
75%              0.000000          1.000000  
max              1.000000          1.000000  

[8 rows x 66 columns]

#### (2)-4. bedcount 
instkind별 평균값을 bedcount의 결측값으로 처리

In [ ]:
train.groupby('instkind')['bedCount'].median()
train['bedCount'] = train['bedCount'].fillna(train.groupby('instkind')['bedCount'].transform('median')) 

train['bedCount'].isnull().sum()

0

In [ ]:
test.groupby('instkind')['bedCount'].median()
test['bedCount'] = test['bedCount'].fillna(test.groupby('instkind')['bedCount'].transform('median')) 
test['bedCount'].isnull().sum()
print(test['bedCount'])

0      300.0
1       44.0
2      276.0
3      363.0
4      213.0
       ...  
121    167.0
122      0.0
123      NaN
124    150.0
126     96.0
Name: bedCount, Length: 123, dtype: float64


**test 파일에서 instkind별 bedCount 중앙값 결측치 처리 -> 123행 결측치 처리 안됨. (이유: traditional_hospital의 bedCount 행 하나밖에 없음.)**

In [ ]:
#결측치 행 추출
test[test['bedCount'].isnull()]

OC      sido    openDate  bedCount              instkind      revenue1  \
123 NaN  gyeonggi  20010530.0       NaN  traditional_hospital  1.833906e+10   

     salescost1          sga1       salary1        noi1  ...  \
123         0.0  1.760117e+10  6.824241e+09  31454472.0  ...   

     liquidLiabilities2  shortLoan2  NCLiabilities2     longLoan2  \
123        1.376742e+09         0.0    1.799886e+09  1.740000e+09   

        netAsset2      surplus2  employee1  employee2  ownerChange  \
123  8.574502e+09  7.136681e+09      132.0      137.0         0.87   

     hospital_count  
123             1.0  

[1 rows x 57 columns]

#### (2)-5. employee1,2 결측치 중앙값 대체

In [ ]:
#employee1 결측치 중앙값 대체 
train['employee1'].median()
train.loc[train['employee1'].isna(),'employee1'] = train['employee1'].median()
train['employee1'].isna().sum()

test['employee1'].median()
test.loc[test['employee1'].isna(),'employee1'] = test['employee1'].median()
test['employee1'].isna().sum()

0

In [ ]:
#employee2 결측치 중앙값 대체 
train['employee2'].median()
train.loc[train['employee2'].isna(),'employee2'] = train['employee2'].median()
train['employee2'].isna().sum()

test['employee2'].median()
test.loc[test['employee2'].isna(),'employee2'] = test['employee2'].median()
test['employee2'].isna().sum()

0

# 3. 파생변수 만들기

### (1) 금융변수

In [36]:
## 성장성
# 매출액 증가율 = 당기매출액/ 전기매출액*100 -100
train['revenueGrowth']=(train['revenue1']/train['revenue2'])*100-100
# 총자산 증가율 = 당기총자산/ 전기총자산*100 -100
train['netAssetGrowth']=(train['netAsset1']+train['debt1'])/(train['netAsset2']+train['debt2'])*100 -100

## 수익성
# 자기자본순이익률(ROE)=당기순이익/자기자본*100
train['ROE']=(train['profit1']/train['netAsset1'])*100
# 매출순이익률 = 당기순이익/매출액*100
train['ROS'] = train.profit1/train.revenue1*100
# 배당성향 = 배당금/당기순이익
train['dividendPayoutRatio'] = (train.profit1-train.surplus1)/train.profit1
# 이자보상비율 = 영업이익/이자비용*100
# (영업손익은 매출액 등의 영업수익으로부터 매출원가ㆍ판매비와관리비 등의 영업비용을 차감한 개념, 영업 외 이익과 비용도 차감)
train['interestCoverageRatio'] = (train.revenue1-train.salescost1-train.sga1-train.noi1-train.noe1)/train.interest1*100

## 안정성
# 유동비율 = 유동자산/유동부채*100 (바람직: 200%이상)
train['liquidRatio']=(train['liquidAsset1']/train['liquidLiabilities1'])*100
# 당좌비율 = 당좌자산/유동부채*100 (바람직: 100%이상)
train['quickRatio']=(train['quickAsset1']/train['liquidLiabilities1'])*100
# 부채비율 = 부채/자본*100
train['debtRatio']=(train['debt1']/(train['debt1']+train['netAsset1']))*100
# 자기자본비율 = 자기자본/총자본*100
train['netAssetRatio']=train.netAsset1/(train.debt1+train.netAsset1)*100

## 활동성
# 재고자산회전율 = 매출원가/평균재고자산
train['inventoryTurnover']=train.salescost1/((train.inventoryAsset1+train.inventoryAsset2)/2)

In [37]:
## 확인
fin_Ratio_list = ['revenueGrowth', 'netAssetGrowth',
    'ROE', 'ROS', 'dividendPayoutRatio', 'interestCoverageRatio',
    'liquidRatio', 'quickRatio', 'debtRatio','netAssetRatio',
    'inventoryTurnover']
train[fin_Ratio_list].describe()
# min값에 0이 없고, max값에 inf가 없다.

revenueGrowth  netAssetGrowth           ROE           ROS  \
count   2.930000e+02    2.930000e+02  2.930000e+02  2.930000e+02   
mean   -4.051014e+06   -3.160462e+06 -7.618191e+03  1.635144e+05   
std     2.668536e+07    1.684379e+07  4.150783e+06  2.730859e+06   
min    -3.724756e+08   -1.702364e+08 -4.265897e+07 -2.254366e+05   
25%    -2.828903e+00   -2.020290e+00 -1.079731e+00 -1.731749e-04   
50%     2.615991e+00    8.564362e-01  2.388571e+00  1.647738e+00   
75%     9.898134e+00    6.642079e+00  8.511832e+00  5.255162e+00   
max     2.666720e+03    1.423927e+02  4.673779e+07  4.673779e+07   

       dividendPayoutRatio  interestCoverageRatio   liquidRatio    quickRatio  \
count           293.000000           2.930000e+02  2.930000e+02  2.930000e+02   
mean            -11.133183           2.823433e+04 -8.247440e+04 -8.134630e+04   
std             187.631190           8.208666e+06  1.414748e+06  1.395363e+06   
min           -2679.774273          -6.185428e+07 -2.421641e+07 -2.388459e+07   
25%               0.998165          -4.148300e+02  4.270190e+01  4.046850e+01   
50%               1.000017          -9.972037e+00  9.722931e+01  9.361960e+01   
75%               1.000106           1.675523e+02  1.826708e+02  1.717671e+02   
max            1397.851412           9.143919e+07  4.954042e+03  4.935403e+03   

        debtRatio  netAssetRatio  inventoryTurnover  
count  293.000000     293.000000       2.930000e+02  
mean    60.160045      39.839955      -1.037958e+04  
std     40.119404      40.119404       8.156192e+04  
min     -0.002944    -437.041102      -1.070380e+06  
25%     46.910296      30.056650      -4.684176e-05  
50%     54.988117      45.011883       7.374375e+00  
75%     69.943350      53.089704       1.851061e+01  
max    537.041102     100.002944       4.187911e+02

### (2) 중소병원/대형병원

In [ ]:
train.loc[train.bedCount>300,'hospital_size']=1
train.loc[train.bedCount<=300,'hospital_size']=0

### (3) 지역 내 병원종류수

In [ ]:
#  이건 모델 돌리고나서 만들기

# 4. 피처 정규화

* 표준화 : 정규분포를 가진 값으로 변환 
            (x-mean)/sd
* 정규화 : 개별 데이터의 크기를 모두 똑같은 단위로 변경
            (x-min)/(max-min)

# 5. 모델 돌리기

## (1) RandomForest

## (2) XGBoost

## (3) LightGBM

In [ ]:
X_features = train.drop(columns = 'OC',axis=1)
y_labels = train['OC']

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X_features, y_labels, test_size=0.2, random_state=156)

In [ ]:
# LightGBM의 파이썬 패키지인 lightgbm에서 LGBMClassifier 임포트
from lightgbm import LGBMClassifier
from sklearn.model_selection import train_test_split
import warnings
warnings.filterwarnings('ignore')

In [ ]:
# 앞서 XGBoost와 동일하게 n_estimators는 400설정
lgbm_wrapper = LGBMClassifier(n_estimators=400)

# LightGBM도 XGBoost와 동일하게 조기중단 수행가능
evals = [(X_test, y_test)]
lgbm_wrapper.fit(X_train, y_train,
                 early_stopping_rounds=100, eval_metric="logloss",
                 eval_set=evals, verbose=True)
preds = lgbm_wrapper.predict(X_test)
pred_proba = lgbm_wrapper.predict_proba(X_test)[:, 1]

In [ ]:
type(train['ownerChange'])

pandas.core.series.Series

In [ ]:
train['ownerChange'].value_counts()

0.87    241
0.13     41
Name: ownerChange, dtype: int64

In [ ]:
train[train['ownerChange'].isnull()]

OC  openDate  bedCount      revenue1    salescost1          sga1  \
48    1  19940425      28.0  8.872427e+09  9.641249e+08  7.638920e+09   
62    1  20170602     128.0  5.236406e+08  2.306773e+07  8.355775e+08   
93    0  20070420     203.0  5.423357e+09  0.000000e+00  5.212041e+09   
206   1  20170120     160.0  3.279139e+09  3.726966e+08  3.507972e+09   
212   0  19990603      66.0  2.233031e+10  8.484657e+08  1.849255e+10   
242   1  20170420       0.0  6.845503e+08  0.000000e+00  5.619356e+08   
263   1  20020123      39.0  3.007769e+09  1.197756e+09  2.258240e+09   
285   1  19881104     119.0  9.618709e+09  0.000000e+00  9.751211e+09   
298   0  19830705      72.5  0.000000e+00  0.000000e+00  5.479428e+07   
299   0  20070101     180.0  2.116892e+09  2.681748e+08  3.286245e+09   

          salary1         noi1          noe1    interest1  ...  sido_sejong  \
48   4.488958e+09   19153521.0  2.160889e+08  204479315.0  ...            0   
62   5.644636e+08    7534205.0  5.263060e+07   52630597.0  ...            0   
93   2.958250e+09   23843435.0  1.288886e+08  128638631.0  ...            0   
206  2.659892e+09    1596568.0  8.306566e+07   83065655.0  ...            0   
212  1.232241e+10  423628901.0  1.304817e+09  417949094.0  ...            0   
242  3.014000e+08    8067090.0  2.748281e+06    1364824.0  ...            0   
263  1.775322e+09  611489427.0  5.279657e+07   52076570.0  ...            0   
285  4.979714e+09  813200615.0  5.543432e+08   27044684.0  ...            0   
298  7.500000e+06  121022160.0  1.379810e+08  137980960.0  ...            0   
299  2.231944e+09  123665065.0  5.366919e+08  514368798.0  ...            0   

     sido_seoul  sido_ulsan  instkind_clinic  instkind_dental_clinic  \
48            1           0                0                       0   
62            0           0                0                       0   
93            0           0                0                       0   
206           0           0                0                       0   
212           0           0                0                       0   
242           0           0                0                       0   
263           0           0                0                       0   
285           0           0                0                       0   
298           0           0                0                       0   
299           0           0                0                       0   

     instkind_general_hospital  instkind_hospital  instkind_nursing_hospital  \
48                           0                  0                          0   
62                           0                  0                          1   
93                           0                  1                          0   
206                          0                  0                          1   
212                          0                  0                          1   
242                          0                  0                          0   
263                          0                  1                          0   
285                          0                  1                          0   
298                          0                  1                          0   
299                          0                  0                          1   

     instkind_traditional_clinic  instkind_traditional_hospital  
48                             0                              1  
62                             0                              0  
93                             0                              0  
206                            0                              0  
212                            0                              0  
242                            1                              0  
263                            0                              0  
285                            0                              0  
298                            0                              0

In [ ]:
train['ownerChange'].isna().sum()

10